pkg/urbit/vere/saga.c

```diff
@@ -0,0 +1,730 @@
+{
+  u3_saga* log_u = c3_calloc(sizeof(*log_u));
+  if ( !(log_u->pax_u = c3_path_fp(pax_u)) ) {
+    goto free_event_log;
+  }
+
+  { // (1)
+    c3_path_push(log_u->pax_u, "data.mdb");
+    c3_i ret_i = access(c3_path_str(log_u->pax_u), R_OK | W_OK);
+    c3_path_pop(log_u->pax_u);
+    if ( 0 == ret_i ) {
+      if ( !_migrate(log_u, met_u) ) {
+        fprintf(stderr,
+                "saga: failed to create first "
+                "epoch from existing event log\r\n");
+        goto free_event_log;
+      }
+      goto succeed;
+    }
+  }
+
+  { // (2)
+    void* dat_v;
+
+    c3_path_push(log_u->pax_u, fak_nam_c);
+    dat_v = &met_u->fak_o;
+    if ( !c3_bile_read_existing(log_u->pax_u, dat_v, sizeof(met_u->fak_o)) ) {
+      fprintf(stderr, "saga: failed to read %s\r\n", c3_path_str(log_u->pax_u));
+      goto free_event_log;
+    }
+    c3_path_pop(log_u->pax_u);
+
+    c3_path_push(log_u->pax_u, who_nam_c);
+    dat_v = met_u->who_d;
+    if ( !c3_bile_read_existing(log_u->pax_u, dat_v, sizeof(met_u->who_d)) ) {
+      fprintf(stderr, "saga: failed to read %s\r\n", c3_path_str(log_u->pax_u));
+      goto free_event_log;
+    }
+    c3_path_pop(log_u->pax_u);
+  }
+
+  c3_c(*ent_c)[dname_size];
+  size_t ent_i;
+  if ( !_read_epoc_dirs(c3_path_str(log_u->pax_u), &ent_c, &ent_i) ) {
+    goto free_event_log;
+  }
+
+  try_list(log_u->epo_u.lis_u = c3_list_init(), goto free_dir_entries);
+  u3_epoc *poc_u, *pre_u;
+  c3_w* lif_w = &met_u->lif_w;
+  for ( size_t idx_i = 0; idx_i < ent_i; idx_i++ ) {
+    c3_path_push(log_u->pax_u, ent_c[idx_i]);
+    try_epoc(poc_u = u3_epoc_open(log_u->pax_u, lif_w), goto free_dir_entries);
+    c3_path_pop(log_u->pax_u);
+    c3_list_pushb(log_u->epo_u.lis_u, poc_u, epo_siz_i);
+    c3_free(poc_u);
+    if ( lif_w ) {
+      lif_w = NULL;
+    }
+  }
+
+  log_u->epo_u.cur_u = c3_lode_data(c3_list_peekb(log_u->epo_u.lis_u));
+  log_u->eve_d       = u3_epoc_last_commit(log_u->epo_u.cur_u);
+
+  try_list(log_u->eve_u.lis_u = c3_list_init(), goto free_dir_entries);
+
+  c3_free(ent_c);
+  goto succeed;
+
+free_dir_entries:
+  c3_free(ent_c);
+free_event_log:
+  u3_saga_close(log_u);
+  c3_free(log_u);
+  return NULL;
+
+succeed:
+  return log_u;
+}
+
+c3_d
+u3_saga_last_commit(const u3_saga* const log_u)
+{
+  return u3_epoc_last_commit(log_u->epo_u.cur_u);
+}
+
+c3_t
+u3_saga_needs_rollover(const u3_saga* const log_u)
+{
+  return u3_epoc_len(log_u->epo_u.cur_u) >= epo_len_i;
+}
+
+//! @n (1) Bootstrap is needed if the only epoch present is the first epoch.
+//! @n (2) Bootstrap is always required on full replay.
+c3_t
+u3_saga_needs_bootstrap(const u3_saga* const log_u)
+{
+#ifndef U3_REPLAY_FULL
+  c3_assert(log_u);
+  const u3_epoc* const poc_u = c3_lode_data(c3_list_peekf(log_u->epo_u.lis_u));
+  const size_t         len_i = c3_list_len(log_u->epo_u.lis_u);
+  return epo_min_d == u3_epoc_first_commit(poc_u) && 1 == len_i; // (1)
+#else
+  return 1; // (2)
+# endif /* ifndef U3_REPLAY_FULL */
```

Comment:
From looking at `_read_epoc_dirs` I imagined it was, and I _think_ all epoch list loading happens through the one callsite for that?

Reply:
@Fang- is right: the epoch list is sorted (and yes, all epoch loading is done through `_read_epoc_dirs()`). This condition checks "is our oldest epoch the first epoch *and* is it the only epoch we have". We only need to bootstrap if both conditions are true. If the former is false, then we can boot from our oldest epoch's snapshot (in the worst case) because it's not the first epoch, which doesn't have a snapshot. If the latter is false, then we can boot from some more recent epoch. This of course is only relevant in the context of truncation, which isn't implemented via the binary yet but does work in the obvious way of `rm -rf <pier>/.urb/log/0i1`.